In [44]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('loan.csv')

/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (47) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   
3                NaN            NaN               NaN                  0.0   
4                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   
3       0.0             NaN               NaN            NaN   
4       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 111 columns]

In [9]:
df.shape

(39717, 111)

In [5]:
#columns present in the dataset
df.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       ...
       'num_tl_90g_dpd_24m', 'num_tl_op_past_12m', 'pct_tl_nvr_dlq',
       'percent_bc_gt_75', 'pub_rec_bankruptcies', 'tax_liens',
       'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit',
       'total_il_high_credit_limit'],
      dtype='object', length=111)

lets check the number of unique loan listing as id is the unique identifier for each listing

In [8]:
df['id'].unique().shape

(39717,)

So it means there are no duplicate loan listing as number number of rows is same

In [10]:
#creating function to get missing values in each column and displaying them
'''
get_summary_df is a function which takes a dataframe and returns
a summary dataframe of number of unique and missing values in each column
'''
def get_summary_df(df):
    print(f"Dataset Shape: {df.shape}")
    df_sum = pd.DataFrame(df.dtypes,columns=['dtypes'])
    df_sum = df_sum.reset_index()
    df_sum['Name'] = df_sum['index']
    df_sum = df_sum[['Name','dtypes']]
    df_sum['Missing'] = df.isnull().sum().values    
    df_sum['Uniques'] = df.nunique().values

    return df_sum

In [12]:
summary_df = get_summary_df(df)

Dataset Shape: (39717, 111)


In [13]:
summary_df

Name   dtypes  Missing  Uniques
0                            id    int64        0    39717
1                     member_id    int64        0    39717
2                     loan_amnt    int64        0      885
3                   funded_amnt    int64        0     1041
4               funded_amnt_inv  float64        0     8205
..                          ...      ...      ...      ...
106                   tax_liens  float64       39        1
107             tot_hi_cred_lim  float64    39717        0
108           total_bal_ex_mort  float64    39717        0
109              total_bc_limit  float64    39717        0
110  total_il_high_credit_limit  float64    39717        0

[111 rows x 4 columns]

lets get the columns which have highest number of missing values

In [15]:
summary_df.sort_values(by='Missing',ascending=False).head()

Name   dtypes  Missing  Uniques
55  verification_status_joint  float64    39717        0
53           annual_inc_joint  float64    39717        0
81       mo_sin_old_rev_tl_op  float64    39717        0
80         mo_sin_old_il_acct  float64    39717        0
77                    bc_util  float64    39717        0

top 5 columns having highest missing values i.e 'verification_status_joint','annual_inc_joint','mo_sin_old_rev_tl_op','mo_sin_old_il_acct','bc_util'

In [16]:
#adding a new column which indicates percentage of missing values
summary_df['missing_perc'] = summary_df['Missing']/39717

In [18]:
summary_df.head()

Name   dtypes  Missing  Uniques  missing_perc
0               id    int64        0    39717           0.0
1        member_id    int64        0    39717           0.0
2        loan_amnt    int64        0      885           0.0
3      funded_amnt    int64        0     1041           0.0
4  funded_amnt_inv  float64        0     8205           0.0

lets get the list of columns whose missing_perc = 1

In [23]:
summary_df[summary_df['missing_perc']==1].Name.count()

54

It means that out of 111 columns 54 columns have no value

In [29]:
removed_cols = summary_df[summary_df['missing_perc']==1].Name.tolist()
tot_cols = df.columns

In [31]:
accepted_cols = [col for col in tot_cols if col not in removed_cols]

In [34]:
#removing those columns from dataframe where there is not even a single value and saving the new dataframe as df_trimmed
df_trimmed = df[accepted_cols]

In [36]:
df_trimmed.shape

(39717, 57)

In [37]:
df_trimmed.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... next_pymnt_d last_credit_pull_d  \
0   10.65%       162.87     B        B2  ...          NaN             May-16   
1   15.27%        59.83     C        C4  ...          NaN             Sep-13   
2   15.96%        84.33     C        C5  ...          NaN             May-16   
3   13.49%       339.31     C        C1  ...          NaN             Apr-16   
4   12.69%        67.79     B        B5  ...       Jun-16             May-16   

  collections_12_mths_ex_med  policy_code application_type acc_now_delinq  \
0                        0.0            1       INDIVIDUAL              0   
1                        0.0            1       INDIVIDUAL              0   
2                        0.0            1       INDIVIDUAL              0   
3                        0.0            1       INDIVIDUAL              0   
4                        0.0            1       INDIVIDUAL              0   

  chargeoff_within_12_mths delinq_amnt pub_rec_bankruptcies tax_liens  
0                      0.0           0                  0.0       0.0  
1                      0.0           0                  0.0       0.0  
2                      0.0           0                  0.0       0.0  
3                      0.0           0                  0.0       0.0  
4                      0.0           0                  0.0       0.0  

[5 rows x 57 columns]

In [40]:
# Lets start exploring some variables to get an overview starting with loan status as it is easiest to interpret

df_trimmed['loan_status'].head()

0     Fully Paid
1    Charged Off
2     Fully Paid
3     Fully Paid
4        Current
Name: loan_status, dtype: object

As we can see loan_status is a categorical variable of nominal type

In [41]:
# lets see what all unique values are present in the variable

df_trimmed['loan_status'].unique()

array(['Fully Paid', 'Charged Off', 'Current'], dtype=object)

In [42]:
# lets check the distribution of different types of loan status

In [49]:
total_count = 39717

plt.figure(figsize = (14,6))
df_plot = df_trimmed
col_name = 'loan_status'
df_plot[col_name+'ratio'] = df_plot[col_name]/total_count
g = plt.plot(,x=col_name+'ratio', data=df_trimmed, 
                  color='blue')
g.set_xticklabels(g.get_xticklabels(),rotation=45)
g.set_xlabel("Loan Status Categories", fontsize=12)
g.set_ylabel("Count", fontsize=15)
g.set_title("Loan Status Types Distribution", fontsize=20)


plt.show()

TypeError: unsupported operand type(s) for /: 'str' and 'int'

<Figure size 1008x432 with 0 Axes>

In [ ]:
g = sns.countplot(x=col_name+'ratio', data=df_plot, 
                  color='blue')
g.set_xticklabels(g.get_xticklabels(),rotation=45)
g.set_xlabel("Loan Status Categories", fontsize=12)
g.set_ylabel("Count", fontsize=15)
g.set_title("Loan Status Types Distribution", fontsize=20)


plt.show()